src/runtime/virtcontainers/container.go

```diff
@@ -862,12 +863,16 @@ func (c *Container) create() (err error) {
 	// lazy attach device after createContainer for q35
 	if machineType == QemuQ35 && len(delayAttachedDevs) > 0 {
 		c.Logger().WithFields(logrus.Fields{
-			"machine_type": machineType,
-			"devices":      delayAttachedDevs,
+			"machine_type":           machineType,
+			"devices":                delayAttachedDevs,
+			"pcie_lazy_attach_delay": delayAttachSeconds,
 		}).Info("lazy attach devices")
-		if err = c.attachDevices(delayAttachedDevs); err != nil {
-			return
-		}
+		// fix: https://github.com/kata-containers/kata-containers/issues/835
+		time.AfterFunc(time.Duration(delayAttachSeconds)*time.Second, func() {
```

Comment:
Dose here need a block or wait? because `time.AfterFunc` will run the function in a new goroutine, and current goroutine will go on to run the next line.

Reply:
This will not block the container from starting.

Reply:
I don't think this is a good idea.  It means the container payload can start before all the devices are attached, which means the payload may or may not work depending on details of the timing.  I think the expectation is that you specify a device to be attached to a container, then when the container entry point executes the device should already be present and usable.

Reply:
Do you mean to change to sync waiting for nvidia gpu device hot plug?
```golang
// fix: https://github.com/kata-containers/kata-containers/issues/835
time.Sleep(time.Duration(delayAttachSeconds)*time.Second)
if err = c.attachDevices(delayAttachedDevs); err != nil {
	return
}
```

- I thought that users can check the nvidia gpu device and load the nvidia driver at its entry point.
In order to avoid slowing down the guest startup, I used time.AfterFunc().
- And these codes only affects nvidia gpu VFIO devices with the big bar space.

Reply:
> Do you mean to change to sync waiting for nvidia gpu device hot plug?
> 
> ```go
> // fix: https://github.com/kata-containers/kata-containers/issues/835
> time.Sleep(time.Duration(delayAttachSeconds)*time.Second)
> if err = c.attachDevices(delayAttachedDevs); err != nil {
> 	return
> }
> ```
> 
>     * I thought that users can check the nvidia gpu device and load the nvidia driver at its entry point.
>       In order to avoid slowing down the guest startup, I used time.AfterFunc().

They can, but requiring the container contents to change in order to accomodate Kata's behaviour means we're not really hitting the goal of being a drop-in replacement for an OCI runtime.

>     * And these codes only affects nvidia gpu VFIO devices with the big bar space.

Ah... your immediate proposal does, but this problem can affect all devices attached using PCI-E native hotplug (in fact the longer term solution I'm tackling for this and some other problems should remove the need for special handling of large bar devices).